src/hotspot/share/opto/replacednodes.cpp

```diff
@@ -106,90 +106,141 @@ void ReplacedNodes::apply(Node* n, uint idx) {
-      while (work.size() != 0 && replace) {
-        Node* n = work.pop();
-        if (use->outcnt() == 0) {
-          continue;
-        }
-        if (n->is_CFG() || (n->in(0) != nullptr && !n->in(0)->is_top())) {
-          // Skip projections, since some of the multi nodes aren't CFG (e.g., LoadStore and SCMemProj).
-          if (n->is_Proj()) {
-            n = n->in(0);
+    while (!stack.is_empty()) {
+      assert(stack.size() > 1, "at least initial + one use");
+      Node* n = stack.node();
+
+      uint current_size = stack.size();
+      if (n->outcnt() != 0 && n != improved) {
+        if (n->is_Phi()) {
+          Node* region = n->in(0);
+          Node* prev = stack.node_at(stack.size() - 2);
+          for (uint j = 1; j < region->req(); ++j) {
+            if (n->in(j) == prev) {
+              Node* in = region->in(j);
+              if (in != nullptr && !in->is_top()) {
+                if (is_dominator(ctl, in)) {
+                  clone_stack(C, initial, improved, stack, j);
+                }
+              }
+            }
           }
-          if (!n->is_CFG()) {
-            n = n->in(0);
+        } else if (n->is_CFG()) {
+          if (is_dominator(ctl, n)) {
+            clone_stack(C, initial, improved, stack, -1);
           }
-          assert(n->is_CFG(), "should be CFG now");
-          int depth = 0;
-          while(n != ctl) {
-            n = IfNode::up_one_dom(n);
-            depth++;
-            // limit search depth
-            if (depth >= 100 || n == nullptr) {
-              replace = false;
-              break;
-            }
+        } else if (n->in(0) != nullptr && !n->in(0)->is_top()) {
+          Node* c = n->in(0);
+          if (is_dominator(ctl, c)) {
+            clone_stack(C, initial, improved, stack, -1);
           }
         } else {
-          for (DUIterator k = n->outs(); n->has_out(k); k++) {
-            enqueue_use(n, n->out(k), work);
+          uint idx = stack.index();
+          if (idx < n->outcnt()) {
+            stack.set_index(idx + 1);
+            stack.push(n->raw_out(idx), 0);
           }
         }
       }
-      if (replace) {
-        bool is_in_table = C->initial_gvn()->hash_delete(use);
-        int replaced = use->replace_edge(initial, improved);
-        if (is_in_table) {
-          C->initial_gvn()->hash_find_insert(use);
+      if (stack.size() == current_size) {
+        for (;;) {
+          stack.pop();
+          if (stack.is_empty()) {
+            break;
+          }
+          n = stack.node();
+          uint idx = stack.index();
+          if (idx < n->outcnt()) {
+            stack.set_index(idx + 1);
+            stack.push(n->raw_out(idx), 0);
+            break;
+          }
         }
-        C->record_for_igvn(use);
-
-        assert(replaced > 0, "inconsistent");
-        --j;
       }
     }
   }
 }
 
+void ReplacedNodes::clone_stack(Compile* C, Node* initial, Node* improved, const Node_Stack& stack, int i) const {
```

Comment:
What about using a more descriptive name here like `clone_uses_and_replace` and maybe add a brief comment?

Reply:
Done in the new commit.